core/kernel/msg_param.c

```diff
@@ -31,6 +31,122 @@
 #include <kernel/msg_param.h>
 #include <mm/mobj.h>
 
+/**
+ * msg_param_extract_pages() - extract list of pages from
+ * OPTEE_MSG_ATTR_NONCONTIG buffer.
+ *
+ * @buffer:	pointer to parameters array
+ * @pages:	output array of page addresses
+ * @num_pages:  number of pages in array
+ *
+ * return:
+ *   page count on success
+ *   <0 on error
+ *
+ * @buffer points to the physical address of a structure that can be viewed as
+ *
+ * struct page_data {
+ *   uint64_t pages_array[OPTEE_MSG_NONCONTIG_PAGE_SIZE/sizeof(uint64_t) - 1];
+ *   uint64_t next_page_data;
+ * };
+ *
+ * So, it is a linked list of arrays, where every element of linked list fits
+ * exactly into one 4K page.
+ *
+ * This function extracts data from arrays into one array pointed by @pages
+ *
+ * @buffer points to data shared with normal world, so some precautions
+ * should be taken.
+ */
+static ssize_t msg_param_extract_pages(paddr_t buffer, paddr_t *pages,
+				       size_t num_pages)
```

Comment:
minor: as noted in a previous comment, this could return a true/false status as it returns either -1 or the value of `num_pages`.